pkg/function/service.go

```diff
@@ -114,6 +116,42 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.
 	return datumList, nil
 }
 
+// MapTFn applies a function to each datum element.
+// In addition to map function, MapTFn also supports assigning a new event time to datum.
+func (fs *Service) MapTFn(ctx context.Context, d *functionpb.Datum) (*functionpb.DatumList, error) {
+	var key string
+
+	// get key from gPRC metadata
+	if grpcMD, ok := grpcmd.FromIncomingContext(ctx); ok {
+		keyValue := grpcMD.Get(DatumKey)
+		if len(keyValue) > 1 {
+			return nil, fmt.Errorf("expect extact one key but got %d keys", len(keyValue))
+		} else if len(keyValue) == 1 {
+			key = keyValue[0]
+		} else {
+			// do nothing: the length equals zero is valid, meaning the key is an empty string ""
+		}
+	}
+	var hd = handlerDatum{
+		value:     d.GetValue(),
+		eventTime: d.GetEventTime().EventTime.AsTime(),
```

Comment:
I am thinking about how we will assign EventTime and Watermark here if we are to use this new user-defined handler to get event-time and then assign watermark.

Reply:
See comments above, I am thinking let user determine whether or not EventTime and Watermark assigned here is valid.